File: android/Reguerta/app/src/main/java/com/reguerta/user/data/firestore/ReguertaFirestorePath.kt
package com.reguerta.user.data.firestore

import com.reguerta.user.BuildConfig
import java.util.concurrent.atomic.AtomicReference

enum class ReguertaFirestoreEnvironment(val wireValue: String) {
    DEVELOP("develop"),
    PRODUCTION("production"),
}

enum class ReguertaFirestoreCollection(val pathComponent: String) {
    USERS("plus-collections/users"),
    PRODUCTS("plus-collections/products"),
    ORDERS("plus-collections/orders"),
    ORDER_LINES("plus-collections/orderlines"),
    SEASONAL_COMMITMENTS("plus-collections/seasonalCommitments"),
    CONFIG("plus-collections/config"),
    DELIVERY_CALENDAR("plus-collections/deliveryCalendar"),
    SHARED_PROFILES("plus-collections/sharedProfiles"),
    SHIFTS("plus-collections/shifts"),
    SHIFT_PLANNING_REQUESTS("plus-collections/shiftPlanningRequests"),
    SHIFT_SWAP_REQUESTS("plus-collections/shiftSwapRequests"),
    NEWS("plus-collections/news"),
    NOTIFICATION_EVENTS("plus-collections/notificationEvents"),
}

enum class ReguertaFirestoreDocument(val wireValue: String) {
    GLOBAL("global"),
}

data class ReguertaFirestorePath(
    val environment: ReguertaFirestoreEnvironment? = null,
) {
    private val resolvedEnvironment: ReguertaFirestoreEnvironment
        get() = environment ?: ReguertaRuntimeEnvironment.currentFirestoreEnvironment()

    fun collectionPath(collection: ReguertaFirestoreCollection): String =
        "${resolvedEnvironment.wireValue}/${collection.pathComponent}"

    fun documentPath(
        collection: ReguertaFirestoreCollection,
        documentId: String,
    ): String = "${collectionPath(collection)}/$documentId"
}

object ReguertaRuntimeEnvironment {
    private val sessionOverride = AtomicReference<ReguertaFirestoreEnvironment?>(null)
    @Volatile
    private var testingBaseEnvironment: ReguertaFirestoreEnvironment? = null

    val baseFirestoreEnvironment: ReguertaFirestoreEnvironment
        get() = testingBaseEnvironment ?: if (BuildConfig.DEBUG) {
            ReguertaFirestoreEnvironment.DEVELOP
        } else {
            ReguertaFirestoreEnvironment.PRODUCTION
        }

    fun currentFirestoreEnvironment(): ReguertaFirestoreEnvironment =
        sessionOverride.get() ?: baseFirestoreEnvironment

    fun applySessionEnvironment(environment: ReguertaFirestoreEnvironment) {
        sessionOverride.set(if (environment == baseFirestoreEnvironment) null else environment)
    }

    fun resetToBaseEnvironment() {
        sessionOverride.set(null)
    }

    internal fun setBaseEnvironmentForTesting(environment: ReguertaFirestoreEnvironment?) {
        testingBaseEnvironment = environment
        resetToBaseEnvironment()
    }
}

File: ios/Reguerta/Reguerta/Data/Firestore/ReguertaFirestorePath.swift
import FirebaseFirestore
import Foundation

typealias ReguertaFirestoreEnvironment = SessionEnvironment

enum ReguertaFirestoreCollection: String, Sendable {
    case users
    case memberDirectory
    case products
    case orders
    case orderlines
    case seasonalCommitments
    case config
    case deliveryCalendar
    case sharedProfiles
    case shifts
    case shiftPlanningRequests
    case shiftPlanningCandidates
    case shiftSwapRequests
    case news
    case notificationEvents

    fileprivate var pathComponent: String {
        "plus-collections/\(rawValue)"
    }
}

enum ReguertaFirestoreDocument: String, Sendable {
    case global
    case memberConfiguration = "member"
    case publicConfiguration = "public"
}

struct ReguertaFirestorePath {
    let environment: ReguertaFirestoreEnvironment

    func collectionPath(_ collection: ReguertaFirestoreCollection) -> String {
        "\(environment.rawValue)/\(collection.pathComponent)"
    }

    func documentPath(in collection: ReguertaFirestoreCollection, documentId: String) -> String {
        "\(collectionPath(collection))/\(documentId)"
    }
}

extension Firestore {
    func reguertaCollection(
        _ firestoreCollection: ReguertaFirestoreCollection,
        environment: ReguertaFirestoreEnvironment
    ) -> CollectionReference {
        self.collection(
            ReguertaFirestorePath(environment: environment).collectionPath(firestoreCollection)
        )
    }

    func reguertaDocument(
        _ firestoreDocument: ReguertaFirestoreDocument,
        in firestoreCollection: ReguertaFirestoreCollection,
        environment: ReguertaFirestoreEnvironment
    ) -> DocumentReference {
        self.document(
            ReguertaFirestorePath(environment: environment)
                .documentPath(in: firestoreCollection, documentId: firestoreDocument.rawValue)
        )
    }
}
